src/flint/types/fmpz_mod.pyx

```diff
@@ -158,13 +184,145 @@ cdef class fmpz_mod(flint_scalar):
+        failure. 
+
+            >>> mod_ctx = fmpz_mod_ctx(163)
+            >>> mod_ctx(2).inverse()
+            fmpz_mod(82, 163)
+            >>> mod_ctx(2).inverse(check=False)
+            fmpz_mod(82, 163)
+        """
+        cdef fmpz_mod res
+        res = fmpz_mod.__new__(fmpz_mod)
+        res.ctx = self.ctx
+
+        if check is False:
+            fmpz_mod_inv(res.val, self.val, self.ctx.val)
+            return res
+
+        cdef bint r
+        cdef fmpz one = fmpz.__new__(fmpz)
+        fmpz_one(one.val)
+
+        r = fmpz_mod_divides(
+            res.val, one.val, self.val, self.ctx.val
+        )
+        if r == 0:
+            raise ZeroDivisionError(f"{self} is not invertible modulo {self.ctx.modulus()}")
+
+        return res
+
+    def discrete_log(self, a, check=False):
+        """
+        Solve the discrete logarithm problem, using `self = g` as a base.
+        Assumes a solution, :math:`a = g^x \pmod p` exists.
+        
+        NOTE: Requires that the context modulus is prime.
+
+        TODO: This could instead be initalised as a class from a 
+        given base and the precomputations could be stored to allow 
+        faster computations for many discrete logs with the same base. 
+
+            >>> F = fmpz_mod_ctx(163)
+            >>> g = F(2)
+            >>> x = 123
+            >>> a = g**123
+            >>> g.discrete_log(a)
+            123
+        """
+        cdef fmpz_mod_discrete_log_pohlig_hellman_t L
+        cdef bint is_prime
+
+        print(f"[DEBUGGING!]")
+        print("\t[DEBUG]: Starting discrete log")
+
+        # Ensure that the modulus is prime
+        if check:
+            is_prime = fmpz_is_probabprime(self.ctx.val.n)
+            if not is_prime:
+                raise ValueError("modulus must be prime")
+
+        print("\t[DEBUG]: Checked Primality")
+    
+        # Then check the type of the input
+        if typecheck(a, fmpz_mod):
+            if self.ctx != (<fmpz_mod>a).ctx:
+                raise ValueError("moduli must match")
+        else:
+            a = self.ctx.any_as_fmpz_mod(a)
+            if a is NotImplemented:
+                raise TypeError
+
+        print("\t[DEBUG]: Converted types")
+        
+        # Initalise the dlog data, all discrete logs are solved with an internally
+        # chosen base `y`
+        fmpz_mod_discrete_log_pohlig_hellman_init(L)
+
+        print("\t[DEBUG]: Initialised L")
+
+        fmpz_mod_discrete_log_pohlig_hellman_precompute_prime(L, self.ctx.val.n)
+
+        print("\t[DEBUG]: Precomputed prime for L")
+
+        # Solve the discrete log for the chosen base and target
+        # g = y^x_g and  a = y^x_a
+        # We want to find x such that a = g^x =>
+        # (y^x_a) = (y^x_g)^x => x = (x_a / x_g) mod (p-1) 
+        cdef fmpz_t x_a
+        cdef fmpz_t x_g
+        # TODO: should this value be stored for efficiency?
+        fmpz_mod_discrete_log_pohlig_hellman_run(x_g, L, self.val)
```

Comment:
Hmmm I would have said yes, so I'm confused about why this is working at all :/

Reply:
It all depends what the uninitialised memory looks like (hence undefined behaviour). Probably the main thing is whether flint thinks that the `fmpz_t` is a small (inline) fmpz or a big (pointer) fmpz. If it is a pointer then I guess that it will try to free the memory or something. If it is an inline fmpz then probably its value is just overwritten.